src/video/mod.rs

```diff
@@ -181,19 +198,142 @@ impl VideoDevice {
 
         InterruptFlags::default()
     }
+}
+
+impl VideoDevice {
+    fn inc_cycle_count(&mut self, cycles: u32) {
+        self.cycle_count += cycles;
+    }
 
-    // Draw a single line
-    fn draw_line(&mut self) {
-        self.renderer.draw_line(self.mem.get_lcd_y(), &mut self.mem, self.cgb_mode);
+    fn frame_cycle_reset(&mut self) {
+        self.cycle_count -= 154 * 456;
+    }
+
+    fn get_cycle_count(&self) -> u32 {
+        self.cycle_count
     }
 }
 
 impl MemDevice for VideoDevice {
     fn read(&self, loc: u16) -> u8 {
-        self.mem.read(loc)
+        match loc {
+            // Raw tile data
+            0x8000..=0x97FF if self.regs.can_access_vram() => {
+                let base = (loc - 0x8000) as usize + (self.vram_bank as usize * 0x1800);
+
+                if base % 2 == 0 {  // Lower bit
+                    self.vram.lock().unwrap().tile_mem.get_pixel_lower_row(base)
+                } else {            // Upper bit
+                    self.vram.lock().unwrap().tile_mem.get_pixel_upper_row(base)
+                }
+            },
+            // Background Map A
+            0x9800..=0x9BFF if self.regs.can_access_vram() => {
+                let index = (loc - 0x9800) as usize;
+                if self.vram_bank == 0 {
+                    self.vram.lock().unwrap().tile_map_0[index]
+                } else {
+                    self.vram.lock().unwrap().tile_attrs_0[index]
+                }
+            },
+            // Background Map B
+            0x9C00..=0x9FFF if self.regs.can_access_vram() => {
+                let index = (loc - 0x9C00) as usize;
+                if self.vram_bank == 0 {
+                    self.vram.lock().unwrap().tile_map_1[index]
+                } else {
+                    self.vram.lock().unwrap().tile_attrs_1[index]
+                }
+            },
+            // Sprite data
+            0xFE00..=0xFE9F if self.regs.can_access_oam() => self.vram.lock().unwrap().object_mem.read(loc - 0xFE00),
+            // Registers
+            0xFF40 => self.regs.read_lcd_control(),
```

Comment:
Could these hardcoded values have a named constant somewhere instead?

Reply:
Maaaaybe, I did consider that at some point. There are definitely places elsewhere which could. However when it comes to memory values I think it just adds more clutter than anything. I think the comments suffice.